units/host_execute.go

```diff
@@ -0,0 +1,157 @@
+	"github.com/evergreen-ci/evergreen/model/event"
+	"github.com/evergreen-ci/evergreen/model/host"
+	"github.com/evergreen-ci/evergreen/util"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/dependency"
+	"github.com/mongodb/amboy/job"
+	"github.com/mongodb/amboy/registry"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+)
+
+const (
+	hostExecuteJobName = "host-execute"
+)
+
+func init() {
+	registry.AddJobType(hostExecuteJobName, func() amboy.Job { return makeHostExecuteJob() })
+}
+
+type hostExecuteJob struct {
+	HostID   string `bson:"host_id" json:"host_id" yaml:"host_id"`
+	Script   string `bson:"script" json:"script" yaml:"script"`
+	job.Base `bson:"job_base" json:"job_base" yaml:"job_base"`
+
+	host *host.Host
+	env  evergreen.Environment
+}
+
+func makeHostExecuteJob() *hostExecuteJob {
+	j := &hostExecuteJob{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    hostExecuteJobName,
+				Version: 0,
+			},
+		},
+	}
+	j.SetDependency(dependency.NewAlways())
+	return j
+}
+
+// NewHostExecuteJob creates a job that executes a script on the host.
+func NewHostExecuteJob(env evergreen.Environment, h host.Host, script string, id string) amboy.Job {
+	j := makeHostExecuteJob()
+	j.env = env
+	j.host = &h
+	j.HostID = h.Id
+	j.Script = script
+	j.SetPriority(1)
+	j.SetID(fmt.Sprintf("%s.%s.%s", hostExecuteJobName, j.HostID, id))
+	return j
+}
+
+func (j *hostExecuteJob) Run(ctx context.Context) {
+	defer j.MarkComplete()
+
+	if err := j.populateIfUnset(); err != nil {
+		grip.Error(message.WrapError(err, message.Fields{
+			"message": "could not populate fields",
+			"host":    j.host.Id,
+			"distro":  j.host.Distro.Id,
+			"job":     j.ID(),
+		}))
+		j.AddError(err)
+		return
+	}
+
+	if j.host.Status != evergreen.HostRunning {
+		grip.Debug(message.Fields{
+			"message": "host is down, not attempting to run script",
+			"host":    j.host.Id,
+			"distro":  j.host.Distro.Id,
+		})
+		return
+	}
+
+	sshOptions, err := j.host.GetSSHOptions(j.env.Settings().Keys[j.host.Distro.SSHKey])
+	if err != nil {
+		grip.Error(message.WrapError(err, message.Fields{
+			"message": "could not get ssh options",
+			"host":    j.host.Id,
+			"distro":  j.host.Distro.Id,
+			"job":     j.ID(),
+		}))
+		j.AddError(err)
+		return
+	}
+
+	hostInfo, err := j.host.GetSSHInfo()
+	if err != nil {
+		grip.Error(message.WrapError(err, message.Fields{
+			"message": "failed to get host information",
+			"host":    j.host.Id,
+			"distro":  j.host.Distro.Id,
+			"job":     j.ID(),
+		}))
+		j.AddError(err)
+	}
+	output := &util.CappedWriter{
+		Buffer:   &bytes.Buffer{},
+		MaxBytes: 1024 * 1024, // 1MB
+	}
+	if err := j.env.JasperManager().CreateCommand(ctx).Host(hostInfo.Hostname).User(hostInfo.User).
+		ExtendRemoteArgs("-p", hostInfo.Port, "-t", "-t").ExtendRemoteArgs(sshOptions...).
+		SetCombinedWriter(output).
+		AppendArgs("bash", "-s", fmt.Sprintf("<<'EOF'\n%s\nEOF", j.Script)).Run(ctx); err != nil {
```

Comment:
as a second commant, should we have a ShellScript() method on command that does bash -s "eof" things?

Reply:
I changed this to just use `ShellScript()`.